modules/nlu/src/backend/pipelines/slots/pre-processor.ts

```diff
@@ -1,54 +1,47 @@
 import * as sdk from 'botpress/sdk'
 
-const SLOTS_REGEX = /\[(.+?)\]\(([\w_\.-]+)\)/gi
-
-type Tag = 'o' | 'B' | 'I'
-
-export interface Token {
-  tag?: Tag
-  value: string
-  slot?: string
-  matchedEntities: string[]
-}
+import { Sequence, Tag, Token } from '../../typings'
 
-export interface Sequence {
-  intent: string
-  cannonical: string
-  tokens: Token[]
-}
+const SLOTS_REGEX = /\[(.+?)\]\(([\w_\.-]+)\)/gi
 
 // TODO replace this for appropriate tokenizer
 const _tokenize = (input: string): string[] => {
   return input.split(' ').filter(w => w.length)
 }
 
-const _generateNonSlotTokens = (input: string): Token[] => {
-  return _tokenize(input).map(
-    t =>
-      ({
-        tag: 'o',
-        value: t,
-        matchedEntities: []
-      } as Token)
-  )
+const _makeToken = (value: string, matchedEntities: string[], start: number, tag = '', slot = ''): Token => {
+  const token = {
+    value,
+    matchedEntities,
+    start,
+    end: start + value.length,
+  } as Token
+
+  if (tag) token.tag = <Tag>tag
+  if (slot) token.slot = slot
+  return token
 }
 
-const _generateSlotTokens = (input: string, slot: string, slotDefinitions: sdk.NLU.IntentSlot[]): Token[] => {
-  const entity = slotDefinitions.find(slotDef => slotDef.name === slot)!.entity
-  return _tokenize(input).map(
-    (t, i) =>
-      ({
-        tag: i === 0 ? 'B' : 'I',
-        value: t,
-        slot,
-        matchedEntities: [entity]
-      } as Token)
-  )
+// TODO use the same algorithm as in the prediction sequence
+const _generateTrainingTokens = (input: string, start: number, slot: string = '', slotDefinitions: sdk.NLU.SlotDefinition[] = []): Token[] => {
+  const matchedEntities = slotDefinitions
+    .filter(slotDef => slot && slotDef.name === slot)
+    .map(slotDef => slotDef.entity)
+
+  return _tokenize(input).map((t, i) => {
+    let tag = 'o'
+    if (slot) tag = i === 0 ? 'B' : 'I'
```

Comment:
1. { }
2. try to avoid single chars variable, it's a bit harder to read. (exception in the for loop). idx or index is good !